### PR TITLE
Media fixes for online mode

### LINKED ISSUE #286: Form's default media references are removed after submission
**Describe the bug**

Found while reproducing #218. When a form has default `jr:` media URLs, those references are no longer resolved after submitting a record.

**To Reproduce**

1. Create a form with default `jr:` media references
2. Submit a new record
3. Reload the form

**Expected behavior**

The default media references will be reloaded.

**Browser and OS (please complete the following information):**

 - OS: macOS
 - Browser: Chrome

**Additional context**

This bug was very likely introduced in #269. I'm already working on a fix.

### PR DESCRIPTION
**Pull request by eyelidlessness**
Fixes #286 and #287.

This is probably not the ideal fix longer term, but being pragmatic, I feel like it's worth the quick win to address such serious bugs.

### CODE REVIEW

**Comment by lognaturel** [comment]
Could you please say a little bit more about what caused the regression and what the possible implications of this change are? `updateMedia` wasn't needed prior to support for `last-saved`, right? What changed?

**Comment by lognaturel** [comment]
Oh, and is there no test to prevent a future regression that would be low-hanging fruit? Or do you just want this merged so you can look at testing separately?

**Comment by eyelidlessness** [comment]
> Could you please say a little bit more about what caused the regression and what the possible implications of this change are? updateMedia wasn't needed prior to support for last-saved, right? What changed?

Sorry, I definitely should have provided more context. The `last-saved` changes introduced [`formCache.init`](https://github.com/enketo/enketo-express/blob/f7565e1cf7bf146f48b2c8d5207c45d3bc8ae296/public/js/src/enketo-webform.js#L48) to online mode, which calls [`formCache.set`](https://github.com/enketo/enketo-express/blob/f7565e1cf7bf146f48b2c8d5207c45d3bc8ae296/public/js/src/module/form-cache.js#L47), which in turn calls [`_swapMediaSrc`](https://github.com/enketo/enketo-express/blob/f7565e1cf7bf146f48b2c8d5207c45d3bc8ae296/public/js/src/module/form-cache.js#L240). This replaces replaces the images' `src` attributes with `data-offline-src` and removes the `src` attribute, in the form's raw HTML string. In offline mode, `updateMedia` then caches those images and updates those same images' `src` attributes with local [`blob:`](https://developer.mozilla.org/en-US/docs/Web/API/Blob) URLs ([constructed here](https://github.com/enketo/enketo-express/blob/f7565e1cf7bf146f48b2c8d5207c45d3bc8ae296/public/js/src/module/file-manager.js#L141)).

> Oh, and is there no test to prevent a future regression that would be low-hanging fruit? Or do you just want this merged so you can look at testing separately?

There are currently no tests of `enketo-webform.js` and I'm sort of intimidated to try, because it intermingles so much DOM and state, and because it initializes on import rather than a function call. On the other hand, having just identified this particular regression I can certainly see why investing in that test coverage would be desirable. And however this fix goes, I think it may be a good idea to backfill tests for all of the top-level `enketo-*` modules.

- - -

There are other potential approaches, which would likely be more easily testable:

- Reconsider how offline cached images are populated (e.g. by removing the intermediate attribute swap and loading them directly). I'd prefer this, but it's a much bigger change.
- Modify `formCache.init` to not call `_swapMediaSrc` in online mode. I think I avoided this because I didn't want to extend online/offline branchy logic deeper into the codebase, but in hindsight it might have been a better approach if only for the test coverage.

**Comment by lognaturel** [comment]
I think I followed? 😅 For now what I'd like to make sure is clear is how much change there's been between 2.7.3 and now. Like is this a fix that re-establishes exactly the same behavior as in 2.7.3 or is it fixing some new behavior? Specifically, am I understanding correctly that the issue is in how media is resolved either from the local db or a remote request? And was this exact same caching present in 2.7.3?

**Comment by eyelidlessness** [comment]
Prior to last-saved:

- Offline mode had the behavior I described, that has not changed
- Online mode had no caching, no attribute swapping, always requested the images directly

In last-saved:

- By virtue of adding the call to `formCache.init`, online mode swapping attributes... but was not caching or loading from the cache as offline mode does

With this fix, in its current state:

- Media loading behavior are the same between offline/online. It occurs to me now that this may not be desirable?

**Comment by lognaturel** [comment]
> It occurs to me now that this may not be desirable?

Right, that's a pretty big change. For some reason I thought the cache was only being used for last-saved and didn't realize the whole caching mechanism was being brought it. The implications there are pretty significant because it means possibly sensitive data file attachments, audio recordings, images that are attached to the form are sitting around forever on some random computer that was used to fill a form once.

**Comment by lognaturel** [comment]
@eyelidlessness have discussed a bit more and agree that the change in caching behavior is not desirable. He will explore 

> Modify formCache.init to not call _swapMediaSrc in online mode.

We'll also eventually try to backfill some tests to provide a safety net around what is actually supposed to be different between online and offline.

**Review by lognaturel — DISMISSED** [review]
Confirmed with @eyelidlessness that this means there's no caching behavior for online-only forms and that he's tried the cases that were previously failing (#287, #286)

File: public/js/src/module/form-cache.js [hunk 1]
```diff
@@ -466,7 +485,13 @@ function _updateCache( survey ) {
...
                     } )
-                    .then( _swapMediaSrc )
+                    .then( formParts => {
+                        if ( !isOnline ) {
+                            return _swapMediaSrc( formParts );
+                        }
+
+                        return formParts;
+                    } )
                     .then( _addBinaryDefaultsAndUpdateModel )
```
Comment (lognaturel) [inline public/js/src/module/form-cache.js:495]: Seeing all these steps makes me think that the survey does get cached even if online. What ensures that it does't?

**Comment by eyelidlessness** [comment]
After some discussion around [this quesiton](https://github.com/enketo/enketo-express/pull/288#discussion_r674426488) we've determined that this is probably not the appropriate solution, and that we should rework some of the changes in #269. Specifically:

- The current behavior is caching forms/surveys in online mode, which may contain sensitive data so that behavior is undesirable
- I'll instead implement a separate store for `last-saved` records. Most of the existing logic will be the same, but it will be separated from `form-cache.js` and Survey logic/storage.
- I'll do a quick spike into whether the storage logic can be implemented together with the rest of the last-saved logic, rather than `store.js`, to avoid spreading its logic across multiple modules.
- As the changes to `enketo-webform.js` in #269 introduced significant regressions (as addressed in this PR as well as in #280), I will also look into backfilling at least some minimal tests of the setup behavior in that module.

**Pull request closed without merging**